### ISSUE TITLE
How to join a krustlet node to a multi-node k8s cluster created with kubeadm

### ISSUE THREAD

**Issue by ZW007** [krustlet/krustlet#664]
Hi! I already have a local k8s cluster (three bare-metal Ubuntu machines, one master, two worker nodes) installed by Kubeadm. Now I want to add a krustlet node (a bare-metal Ubuntu machine) to the cluster. This [guidance](https://docs.krustlet.dev/howto/bootstrapping/) gives three steps (1) $ ./scripts/bootstrap.sh  (2) $ KUBECONFIG=~/.krustlet/config/kubeconfig krustlet-wasi --port 3000 --bootstrap-file /path/to/your/bootstrap.conf  After these two,  Approve by (3) $ kubectl certificate approve <hostname>-tls  But where should 1.2.3. steps run on? master server or worker server? 

If worker node, but does that mean I should install kubectl on the worker node first? And how can `kubectl` talks to a node that is not master? Doing kubectl on a new node does not make sense to me, also has error: `The connection to the server localhost:8080 was refused - did you specify the right host or port?`

Also, if the worker node server that I want to join as a `Krustlet` worker node is already a k8s `Kubelet` worker node. Should I use a different server, or disjoin the `Kubelet` node from the cluster first before joining the `Krustlet` node to the cluster?

**Comment by ZW007** [comment]
I also do not understand the first part of step (2) `$ KUBECONFIG=~/.krustlet/config/kubeconfig`  There is no `kubeconfig` file under `~/.krustlet/config` folder. There is only a `bootstrap.conf` file in this folder after running (1)

**Comment by bacongobbler** [comment]
>  But where should 1.2.3. steps run on? master server or worker server?

You should run those commands from wherever you plan to run a new kubernetes node. Given your two choices I'd assume this is the "worker" server.

> If worker node, but does that mean I should install kubectl on the worker node first? 

Wherever krustlet is running, it will require to establish a connection to the API server. The node bootstrapping process will generate the kube config.

Krustlet contains an embedded kubernetes client, so you should not need to have kubectl installed on the worker node.

> I also do not understand the first part of step (2)

During the node bootstrapping process, a kubeconfig file is generated and placed in the location where you set your KUBECONFIG environment variable. This is described [in the documentation](https://docs.krustlet.dev/howto/bootstrapping/#running-krustlet):

> A couple important notes here. KUBECONFIG should almost always be set, especially in developer/local machine situations. During the bootstrap process, Krustlet will generate a kubeconfig with the credentials it obtains during the bootstrapping process and write it out to the specified location in KUBECONFIG. If a kubeconfig already exists there, it will be loaded and skip the bootstrapping process. A similar process occurs during the bootstrapping of the serving certificates, they will be written out to the paths specified by --cert-file (default $KRUSTLET_DATA_DIR/config/krustlet.crt) and --private-key-file (default $KRUSTLET_DATA_DIR/config/krustlet.key). If they already exist, then they will be loaded and bootstrapping skipped.

Let me know if you'd like me to clarify something.

> Should I add something like `--node-ip 172.17.0.1` when i run `krustlet-wasi`?

Your `--node-ip` flag should be whatever IP address the Kubernetes API server uses to issue API requests to the node. For example when you issue the command `kubectl logs my-pod`, your kubectl client will issue a request against the Kubernetes API server. Once established, the API server will look for the node running pod `my-pod`. The API server will then issue a `GET /logs` request against the kubelet (krustlet) to fetch the logs.

in other words, if - from the worker node - 172.17.0.1 is the IP address that can be used to reach the API server, then yes.

More information can be found in the HOWTO guides: https://docs.krustlet.dev/howto/krustlet-on-kind/#step-2-determine-the-default-gateway

Hope this helps.

**Comment by ZW007** [comment]
@bacongobbler Thank you Matthew. After copying `$HOME/.kube/config` file from the master server to the worker server, my worker node's `kubectl` is now configured and I can run `kubectl` on a worker node now.  I can run 1)2)3) on a worker node. 

After running 1) `./bootstrap.sh` successfully on a worker node. However there is no `kubeconfig` file generated in `~/.krustlet/config` folder. There is only a `bootstrap.conf` file in this folder after running (1). 

so `KUBECONFIG=~/.krustlet/config/kubeconfig` does not make sense to me though I did so. And then I run 2) `krustlet-wasi --node-ip 10.177.48.142 --bootstrap-file ~/.krustlet/config/bootstrap.conf` where node-ip is the internal IP of my worker node. I then run `kubectl certificate approve u4067-tls` on another termimal of this worker node. 

I did see my `krustlet` node ready when use `kubectl get no`, but I failed to try this `kubectl apply -f greet.yaml`on the kubectl node. The `greet` pod is not ready. I did `az login` on the worker  node terminal first before apply

`NAMESPACE      NAME                                  READY     STATUS       RESTARTS   AGE`
`default                 greet                                     0/1             ExitCode:0   0          6m39s`

**Comment by bacongobbler** [comment]
Looks to me like the application succeeded. Notice that the status output shows the state as "Terminated" and the message was "Module run completed". If you run `kubectl logs greet` you should see the log output from the module's execution.

**Comment by ZW007** [comment]
> Looks to me like the application succeeded. Notice that the status output shows the state as "Terminated" and the message was "Module run completed" with an exit code of zero, which means it ran successfully. If you run `kubectl logs greet` you should see the log output from the module's execution.

yes,  `kubectl logs greet` did give me `hello world` though pod status showed not ready. More weirdly. After deleting the pod and doing `apply -f` again, `kubectl logs greet` gave me error, and `krustlet` tells me authentication required for which I do not know how. I guess the problem is about with the registry authentication? 

![image](https://user-images.githubusercontent.com/45562036/129666724-c12f7809-f102-49b9-a8fc-4cff63121925.png)

![image](https://user-images.githubusercontent.com/45562036/129666802-6928071b-e967-4c25-8c36-fa4ed681693d.png)

**Comment by ZW007** [comment]
@bacongobbler Please ignore my last reply. I retried and 'az login' and restarted `krustlet`. Now `kubectl logs greet` works again. But I am still curious why the pod status not ready and I cannot access the pod through IP address. This means I cannot call it via http request. 

![image](https://user-images.githubusercontent.com/45562036/129778480-ea96dfdf-f12e-4148-aab8-4cac178a15d8.png)

**Comment by bacongobbler** [comment]
The use cases for WebAssembly differ from linux containers. In a container you can run long-standing servers listening for TCP requests, like a web server.

The greet example you're running in your cluster prints a bunch of text to standard output and exits thereafter. Hence why it appears to be "not ready", when it actuality it executed, ran to completion, then exited successfully.

Until WebAssembly can acquire a TCP socket connection via `sock_send`/`sock_recv`, we can only run simple workloads like the greet example. Other providers like [the wasmCloud provider](https://github.com/wasmCloud/krustlet-wasmcloud-provider) work around limitations in the WebAssembly spec by providing an actor/capability programming model so that Krustlet can handle incoming requests, pass the request into the WebAssembly runtime, and parse its response back to the client.

**Comment by bacongobbler** [comment]
As for your helloworld-c issues... Please see https://github.com/krustlet/krustlet/issues/635.

**Comment by ZW007** [comment]
Thank you @bacongobbler  So I guess this [demo](https://www.youtube.com/watch?v=pEEdJlH-nWo) is based on [an old version](URL) of krustlet I guess. If I would like to have my wasm pod listening to the HTTP request, I guess I have to choose the older version of your release, right?

**Comment by bacongobbler** [comment]
No. That demonstration is with the wasmCloud provider. Formerly known as the waSCC provider. You’ll notice the demo refers to a “demos/wascc” directory which are now available in the wasmCloud provider.

https://github.com/wasmCloud/krustlet-wasmcloud-provider
